src/kcas.ml

```diff
@@ -10,69 +10,102 @@ module Id = struct
-    (status : [ `Before | `After ]) =
+      release_before casn lt |> ignore;
+      if not (is_cmp casn state) then (
+        state.after <- state.before;
+        state.casn <- casn_before);
+      release_before casn gt
+
+let release casn cass = function
+  | `After -> release_after casn cass
+  | `Before -> release_before casn cass
+
+let rec verify casn = function
+  | NIL -> `After
+  | CASN (atom, desired, lt, gt) -> (
+      match verify casn lt with
+      | `After ->
+          if is_cmp casn desired && Atomic.get atom.state != desired then
+            `Before
+          else verify casn gt
+      | status -> status)
+
+let finish casn (`Undetermined cass as undetermined) (status : determined) =
   if Atomic.compare_and_set casn (undetermined :> status) (status :> status)
-  then release cass status
+  then release casn cass status
   else Atomic.get casn == `After
 
-let rec determine casn undetermined skip = function
-  | NIL -> skip || finish casn undetermined `After
-  | CASN (loc, desired, lt, gt) as eq ->
-      determine casn undetermined true lt
-      &&
-      let current = Atomic.get loc.state in
-      if desired == current then determine casn undetermined skip gt
-      else
-        let expected = desired.before in
-        if
-          expected == current.after
-          && (current.casn == casn_after || is_after current.casn)
-          || expected == current.before
-             && (current.casn == casn_before || not (is_after current.casn))
-        then
-          let status = Atomic.get casn in
-          if status != (undetermined :> status) then status == `After
-          else if Atomic.compare_and_set loc.state current desired then
-            determine casn undetermined skip gt
-          else determine casn undetermined skip eq
-        else finish casn undetermined `Before
+let exit _ = raise Exit [@@inline never]
+
+let rec determine casn action = function
+  | NIL -> action
+  | CASN (loc, desired, lt, gt) as eq -> (
+      match determine casn action lt with
+      | `Before -> `Before
+      | action ->
+          let current = Atomic.get loc.state in
+          if desired == current then
+            let action = if is_cmp casn desired then `Verify else action in
+            determine casn action gt
+          else
+            let expected = desired.before in
+            if
+              (not (is_cmp casn desired))
+              && (expected == current.after
+                  && (current.casn == casn_after || is_after current.casn)
+                 || expected == current.before
+                    && (current.casn == casn_before
+                       || not (is_after current.casn)))
+            then
+              match Atomic.get casn with
+              | `Undetermined _ ->
+                  if Atomic.compare_and_set loc.state current desired then
+                    determine casn action gt
+                  else determine casn action eq
+              | #determined -> exit ()
```

Comment:
Why raise rather than return observed outcome?

Reply:
The reason for this is to keep the conditional that determines whether to continue the tree traversal essentially as a single comparison:

```ocaml
      match determine casn action lt with
      | `Before -> `Before
      | action ->
```

This is one of the spots in the implementation that took me numerous iterations.  The goal is to avoid extra overhead from the choice of two modes and the approach in the PR is the best (lowest overhead) I could find.

The previous algorithm, which only had a lock-free mode, is simpler as it only returns a boolean outcome.  The new version cannot be implemented so simply, because there are more outcomes.

In the new algorithm, the result of `determine` is either `Before` (potential failure), `After` (potential success and no verification required), `Verify` (potential success requiring verification of compare operations), or `exception Exit` indicating that the `casn` was already determined by some other domain.  Separating all of these cases efficiently is necessary for best performance.